### PR TITLE
[BEAM-2515] BeamSql: refactor the MockedBeamSqlTable and related tests

### PR DESCRIPTION
**Pull request by xumingming**
Summary:

* rename `MockedBeamSqlTable` to `MockedBoundedTable`
* Use the `RowsBuilder` to build rows in Unit Test.
* use `PAssert` rather than `assertEquals` when test PCollections.

### CODE REVIEW

**Comment by xumingming** [comment]
R: @XuMingmin 
R: @takidau

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/TestUtils.java [hunk 1]
```diff
@@ -81,36 +80,20 @@ public void processElement(ProcessContext ctx) {
...
-      int fieldCount = type.size();
-      for (int i = 0; i < args.length; i += fieldCount) {
-        BeamSqlRow row = new BeamSqlRow(type);
-        for (int j = 0; j < fieldCount; j++) {
-          row.addField(j, args[i + j]);
-        }
-        this.rows.add(row);
-      }
-
+    public RowsBuilder addRows(final Object... args) {
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/TestUtils.java]: Much better name, thanks!

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedBoundedTable.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+
+/**
+ * Mocked table for bounded data sources.
+ */
+public class MockedBoundedTable extends MockedTable {
+  public static final ConcurrentLinkedQueue<BeamSqlRow> CONTENT = new ConcurrentLinkedQueue<>();
+  private List<BeamSqlRow> rows = new ArrayList<>();
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedBoundedTable.java]: final

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedBoundedTable.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+   *       Types.INTEGER, "site_id",
+   *       Types.DOUBLE, "price",
+   *       Types.TIMESTAMP, "order_time")
+   * }</pre>
+   */
+  public static MockedBoundedTable of(final Object... args){
+    return new MockedBoundedTable(buildBeamSqlRecordType(args));
+  }
+
+  public MockedBoundedTable addRows(Object... args) {
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedBoundedTable.java]: Javadoc

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedBoundedTable.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+
+/**
+ * Mocked table for bounded data sources.
+ */
+public class MockedBoundedTable extends MockedTable {
+  public static final ConcurrentLinkedQueue<BeamSqlRow> CONTENT = new ConcurrentLinkedQueue<>();
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedBoundedTable.java]: Javadoc

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedUnboundedTable.java [hunk 5]
```diff
@@ -38,14 +39,16 @@
  * A mocked unbounded table.
  */
 public class MockedUnboundedTable extends MockedTable {
-  private List<Pair<Duration, List<BeamSqlRow>>> timestampedRows = new ArrayList<>();
+  /** rows flow out from this table with the specified watermark instant. */
+  private final List<Pair<Duration, List<BeamSqlRow>>> timestampedRows = new ArrayList<>();
```
Comment (XuMingmin) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedUnboundedTable.java]: for L102-L105, is it better to add data first, and then advance watermark?

Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedUnboundedTable.java]: Actually I did it in purpose, Let's look at some examples. First example:

```java
     addRows(
            Duration.ZERO,
            1, 1, 1, FIRST_DATE,
            1, 2, 2, FIRST_DATE
        )
```

It means these data are added after the specified watermark(ZERO) -- just after the first window starts. Another example:

```java
        .addRows(
            WINDOW_SIZE.minus(Duration.standardSeconds(1)),
            2, 2, 3, SECOND_DATE,
            2, 3, 3, SECOND_DATE
        )
```

It means these data are emitted just before the first window close. One more example:


```java
        .addRows(
            WINDOW_SIZE.plus(Duration.standardSeconds(1)),
            2, 2, 3, SECOND_DATE,
            2, 3, 3, SECOND_DATE
        )
```

It means these data are emitted just after the first window close, and after the second window starts.

Very handy and easy to use, right :)

**Comment by xumingming** [comment]
@XuMingmin For the test data set, sometimes we need to test the normal case -- the test would pass; sometimes we need to test the exception case -- the test need to fail, the test data for these two kinds of case will differ.

Another perspective: If we want to use one common test data set to test every cases of every relational algebra(join, union, values etc), the test data set will need diversity. Change to the common data set will affect all of tests, it will make the tests brittle -- which will bring unnecessary maintainance burden.

I prefer prepare separate data set for each test.

**Comment by xumingming** [comment]
Retest this please.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedUnboundedTable.java (hunk 5 above)
Comment (XuMingmin) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/mock/MockedUnboundedTable.java]: hmm, interesting design!

**Comment by XuMingmin** [comment]
Make sense for different functions like JOIN/UNION/..., would be helpful to separate the `prepare()` part which can be leveraged to test DSL methods.

**Comment by xumingming** [comment]
I agree, the data set can be reused to test DSL methods, but let's leave to DSL Unit Tests PR, when it is needed?

**Comment by xumingming** [comment]
@lukecwik can you help merge this one? @XuMingmin and @takidau have already reviewed and approved this PR, just need a merge. (Other ISSUE is blocked by this one).

**Comment by lukecwik** [comment]
Merged, please close.

**Pull request closed without merging**